[quant] Add intrinsic module mappings

jerryzh168 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#23753 [quant] Add intrinsic module mappings**

Summary:
Add intrinsic(fused) module mappings in quantize.py to enable mapping fused modules
in both QAT and post PTQ

Test Plan:
```
python test/test_qat.py
python test/test_quantization.py
```

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

Differential Revision: [D16820749](https://our.internmc.facebook.com/intern/diff/D16820749)

jerryzh168 committed:
[quant] Add intrinsic module mappings

Summary:
Add intrinsic(fused) module mappings in quantize.py to enable mapping fused modules
in both QAT and post PTQ

Test Plan:
```
python test/test_qat.py
python test/test_quantization.py
```

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

raghuramank100 reviewed (changes requested):
Can you please add a test that checks for module swapping for the newly added entries in the default_module_mapping table?

jerryzh168 committed:
Update on "[quant] Add intrinsic module mappings"


Summary:
Add intrinsic(fused) module mappings in quantize.py to enable mapping fused modules
in both QAT and post PTQ

Test Plan:
```
python test/test_qat.py
python test/test_quantization.py
```

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

jerryzh168 committed:
Update on "[quant] Add intrinsic module mappings"


Summary:
Add intrinsic(fused) module mappings in quantize.py to enable mapping fused modules
in both QAT and post PTQ

Test Plan:
```
python test/test_qat.py
python test/test_quantization.py
```

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

jerryzh168 committed:
Update on "[quant] Add intrinsic module mappings"


Summary:
Add intrinsic(fused) module mappings in quantize.py to enable mapping fused modules
in both QAT and post PTQ

Test Plan:
```
python test/test_qat.py
python test/test_quantization.py
```

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

jerryzh168 committed:
Update on "[quant] Add intrinsic module mappings"


Summary:
Add intrinsic(fused) module mappings in quantize.py to enable mapping fused modules
in both QAT and post PTQ

Test Plan:
```
python test/test_qat.py
python test/test_quantization.py
```

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

raghuramank100 commented on test/common_quantization.py:
```diff
@@ -51,7 +52,6 @@ def test_only_train_fn(model, train_data, loss_fn=_default_loss_fn):
             optimizer.zero_grad()
             output = model(data)
             loss = loss_fn(output, target)
-            loss.backward()
```
Why is this line removed? Without it backprop will not occur.

jerryzh168 replied:
oh, removed when debugging, will add back

raghuramank100 commented on test/test_qat.py:
```diff
@@ -129,9 +129,8 @@ def ref_op(x):
-            result_actual.backward(dout, retain_graph=True)
-            grad_ref = input.grad.cpu()
             result_actual.backward(dout)
+            grad_ref = input.grad.cpu()
```
Gradient calculation for the reference is missing. Suggest:
```python
result_ref = ref_op(input)
result_actual = qat_op(input)
dout = torch.randn(result_ref.size(), dtype=torch.float)
result_ref.backward(dout)
grad_ref = input.grad.cpu()
input.grad.data.zero_()
result_actual.backward(dout)
grad_actual = input.grad.cpu()
self.assertEqual(grad_ref, grad_actual)
```

w1.grad.data.zero_()
result_ref.backward(dout)

jerryzh168 committed:
Update on "[quant] Add intrinsic module mappings"


Summary:
Add intrinsic(fused) module mappings in quantize.py to enable mapping fused modules
in both QAT and post PTQ

Test Plan:
```
python test/test_qat.py
python test/test_quantization.py
```

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

jerryzh168 committed:
Update on "[quant] Add intrinsic module mappings"


Summary:
Add intrinsic(fused) module mappings in quantize.py to enable mapping fused modules
in both QAT and post PTQ

Test Plan:
```
python test/test_qat.py
python test/test_quantization.py
```

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

zafartahirov commented on test/common_quantization.py:
```diff
@@ -90,7 +91,7 @@ def checkObservers(self, module):
         """
         if hasattr(module, 'qconfig') and module.qconfig is not None and len(module._modules) == 0:
-            self.assertTrue(hasattr(module, 'observer'))
+            self.assertTrue(hasattr(module, 'observer'), 'module: ' + str(type(module)) + ' do not have observer')
```
80 char limit?

```python
self.assertTrue(hasattr(module, 'observer'), \
    'module: ' + str(type(module)) + ' do not have observer')
```

zafartahirov commented on test/common_quantization.py:
```diff
@@ -352,30 +353,50 @@ def forward(self, x):
-        self.conv = torch.nn.Conv2d(20, 20, 1, bias=None)
-        self.bn = torch.nn.BatchNorm2d(20)
+        super(SubModelForFusion, self).__init__()
+        self.conv = nn.Conv2d(2, 2, 1, bias=None).to(dtype=torch.float)
```
Why do we need to convert the layer to `float`? I think by default it assumes so.

jerryzh168 replied:
by default it is torch.double

zafartahirov commented on torch/nn/_intrinsic/qat/modules/conv_fused.py:
```diff
@@ -150,11 +148,11 @@ def from_float(cls, mod, qconfig=None):
         if not qconfig:
             assert hasattr(mod, 'qconfig'), 'Input float module must have qconfig defined'
-            assert hasattr(mod, 'observer'), 'Input float module must have observer attached'
+            assert mod.qconfig, 'Input float module must has valid qconfig'
```
nit: `have valid qconfig`

zafartahirov commented on torch/nn/_intrinsic/quantized/modules/conv_relu.py:
```diff
@@ -22,6 +24,18 @@ def __init__(self, in_channels, out_channels, kernel_size, stride=1,
                                          groups=groups, bias=bias, padding_mode=padding_mode)
 
+    @property
+    def weight(self):
```
Did we decide to keep properties for the weight?

jerryzh168 replied:
no, these should be removed

jerryzh168 replied:
this pr already has too many changes, will do it in a separate pr

raghuramank100 replied:
If you are inheriting from nn.quantized.conv2d, with the latest changes, you dont need these functions anymore, delete these lines?

jerryzh168 committed:
Update on "[quant] Add intrinsic module mappings"


Summary:
Add intrinsic(fused) module mappings in quantize.py to enable mapping fused modules
in both QAT and post PTQ

Test Plan:
```
python test/test_qat.py
python test/test_quantization.py
```

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

zafartahirov commented on torch/nn/utils/fusion.py:
```diff
@@ -8,22 +8,16 @@ def fuse_conv_bn_eval(conv, bn):
+def fuse_conv_bn_weights(conv_w, conv_b, bn_rm, bn_rv, bn_eps, bn_w, bn_b):
+    if conv_b is None:
+        conv_b = bn_rm.new_zeros(bn_rm.shape)
+    bn_var_sqrt = torch.sqrt(bn_rv + bn_eps)
```
nit: `bn_var_sqrt = torch.rsqrt(bn_rv + bn_eps)`

zafartahirov commented on torch/nn/utils/fusion.py:
```diff
@@ -8,22 +8,16 @@ def fuse_conv_bn_eval(conv, bn):
-    fused_conv.weight = torch.nn.Parameter(w_conv)
-    fused_conv.bias = torch.nn.Parameter(b_conv)
+    conv_w = conv_w * (bn_w / bn_var_sqrt).reshape([-1, 1, 1, 1])
+    conv_b = (conv_b - bn_rm) / bn_var_sqrt * bn_w + bn_b
```
nit: If using `rsqrt` above, you can rewrite to

```python
    conv_w = conv_w * (bn_w * bn_var_rsqrt).reshape([-1, 1, 1, 1])
    conv_b = (conv_b - bn_rm) * bn_var_rsqrt * bn_w + bn_b
```

z-a-f reviewed (approved):
Minor nits

jerryzh168 committed:
Update on "[quant] Add intrinsic module mappings"


Summary:
Add intrinsic(fused) module mappings in quantize.py to enable mapping fused modules
in both QAT and post PTQ

Test Plan:
```
python test/test_qat.py
python test/test_quantization.py
```

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

jerryzh168 committed:
Update on "[quant] Add intrinsic module mappings"


Summary:
Add intrinsic(fused) module mappings in quantize.py to enable mapping fused modules
in both QAT and post PTQ

Test Plan:
```
python test/test_qat.py
python test/test_quantization.py
```

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

ezyang commented:
@jerryzh168 Hi Jerry, I'm not too familiar with the workstream here. What kind of review are you looking for?

jerryzh168 commented on torch/nn/_intrinsic/qat/modules/conv_fused.py:
```diff
@@ -129,8 +127,8 @@ def _forward(self, input):
 
-            self.running_mean = exponential_average_factor * batch_mean + (1 - exponential_average_factor) * self.running_mean
-            self.running_var = exponential_average_factor * batch_var + (1 - exponential_average_factor) * self.running_var
+            self.running_mean = exponential_average_factor * batch_mean.detach() + (1 - exponential_average_factor) * self.running_mean
```
@ezyang just want to make sure these detach() are used correctly.

ezyang replied:
LGTM

raghuramank100 commented on test/test_qat.py:
```diff
@@ -123,17 +122,21 @@ def ref_op(x):
-            result_actual = qat_op(input)
-            self.assertEqual(result_ref, result_actual)
+            input_clone = input.clone().detach_().requires_grad_()
+            for i in range(2):
```
This looks like a typo. Do you need the for loop here?

jerryzh168 replied:
yeah, I had some error previously when I run the op with forward and backward twice

raghuramank100 commented on test/test_qat.py:
```diff
@@ -123,17 +122,21 @@ def ref_op(x):
+                grad_ref = input.grad.cpu()
+                loss = (result_actual - dout).sum()
+                loss.backward()
+                grad_actual = input_clone.grad.cpu()
```
NIT: You dont need to create input_clone, you could set gradient to zero after you calculate grad_ref.
```
input.grad.zero_()
```
would do that.

jerryzh168 replied:
in that case we still need to store the gradient of input somewhere before we run a different graph.

raghuramank100 commented on torch/nn/qat/modules/linear.py:
```diff
@@ -39,11 +39,11 @@ def from_float(cls, mod, qconfig=None):
         if not qconfig:
             assert hasattr(mod, 'qconfig'), 'Input float module must have qconfig defined'
-            assert mod.qconfig, 'Input float module must has valid qconfig'
+            assert mod.qconfig, 'Input float module must have a valid qconfig'
```
Do we need this check after  assert hasattr(mod, 'qconfig'). Is the second check ensuring that qconfig is not None?

jerryzh168 replied:
yeah this is to make sure we have a valid qconfig

raghuramank100 commented on torch/nn/_intrinsic/quantized/modules/conv_relu.py:
```diff
@@ -33,3 +48,11 @@ def forward(self, input):
+            mod.weight, mod.bias = \
+                fuse_conv_bn_weights(mod.weight, mod.bias, mod.running_mean,
+                                     mod.running_var, mod.eps, mod.gamma, mod.beta)
+        return super(ConvReLU2d, cls).from_float(mod)
```
Wouldn't super return a nn.quantized.conv2d module instead of a nn._intrinsic.quantized.conv2drelu?

jerryzh168 replied:
no, we are using cls in the from_float

raghuramank100 replied:
I see, cls is pointing to conv2drelu

raghuramank100 commented on torch/nn/quantized/modules/conv.py:
```diff
@@ -126,14 +129,21 @@ def from_float(cls, mod):
+            # workaround for sequential, ConvReLU2d should probably
+            # inherit from Conv2d instead
+            if type(mod) == nni.ConvReLU2d:
+                mod = mod[0]
```
We have a problem when we deal with nn._intrinsic.convReLU2d. We want to use the observer after the ReLU, but here we end up using the observer attached to the conv.

raghuramank100 commented on torch/nn/quantized/modules/linear.py:
```diff
@@ -196,10 +199,15 @@ def from_float(mod):
+            # workaround for sequential, ConvReLU2d should probably
+            # inherit from Conv2d instead
+            if type(mod) == nni.LinearReLU:
+                mod = mod[0]
```
Similar comments for linear

raghuramank100 commented on torch/quantization/quantize.py:
```diff
@@ -19,7 +24,9 @@ def propagate_qconfig_helper(module, qconfig_dict, qconfig_parent=None, prefix='
     """
-    if not hasattr(module, 'qconfig'):
+    if type(module) in skip_list:
+        assert not hasattr(module, 'qconfig'), str(type(module)) + ' module is in skip list, it should not have qconfig'
```
This is not a good user experience. Can we ignore qconfig for modules in skip list?

jerryzh168 replied:
you mean silently skip these modules?

raghuramank100 commented on torch/quantization/quantize.py:
```diff
@@ -240,6 +258,11 @@ def convert(module, mapping=DEFAULT_MODULE_MAPPING):
     reassign = {}
+    # TODO(jerryzh): remove after deciding on the impl of
+    # intrinsic moudles
+    if type(module) in [nni.ConvBn2d, nni.ConvBnReLU2d]:
```
Why only these two modules? What about nni.ConvRelu2d and nni.LinearRelu2d?

jerryzh168 replied:
yeah we need these as well

jerryzh168 committed:
Update on "[quant] Add intrinsic module mappings"


Summary:
Add intrinsic(fused) module mappings in quantize.py to enable mapping fused modules
in both QAT and post PTQ

Test Plan:
```
python test/test_qat.py
python test/test_quantization.py
```

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

raghuramank100 commented on test/test_qat.py:
```diff
@@ -123,17 +122,21 @@ def ref_op(x):
+                loss = (result_actual - dout).sum()
+                loss.backward()
+                grad_actual = input_clone.grad.cpu()
+                self.assertEqual(grad_ref, grad_actual)
```
Please check gradients for all parameters in addition to the input: weight, bias, gamma, beta. 
In addition, please check if all running mean/variance estimates exactly match after forward.

jerryzh168 committed:
Update on "[quant] Add intrinsic module mappings"


Summary:
Add intrinsic(fused) module mappings in quantize.py to enable mapping fused modules
in both QAT and post PTQ

Test Plan:
```
python test/test_qat.py
python test/test_quantization.py
```

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

raghuramank100 commented on test/test_qat.py:
```diff
@@ -123,17 +122,40 @@ def ref_op(x):
+                self.assertEqual(beta_grad_ref, beta_grad_actual)
+                self.assertEqual(num_batches_tracked_ref, num_batches_tracked_actual)
+                self.assertEqual(running_mean_ref, running_mean_actual)
+                # self.assertEqual(running_var_ref, running_var_actual, prec=1e-2)
```
The problem is due to incorrect calculation of variance. Batch norm code at: https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/Normalization.cpp#L223 uses unbiased variance estimate, while the convbn code does not:

https://github.com/pytorch/pytorch/blob/3c5f7c2e4656c84a79168be4e0b7edd8d41e5e2e/torch/nn/_intrinsic/qat/modules/conv_fused.py#L122
Please change:
```
          batch_var = torch.var(conv_orig, dim=[0, 2, 3], unbiased=False)
```
To:
```
          batch_var = torch.var(conv_orig, dim=[0, 2, 3], unbiased=True)
```
  and check if the mismatch in running_var is gone.

raghuramank100 reviewed (changes requested):
Some suggestions to get running_var to match

jerryzh168 committed:
Update on "[quant] Add intrinsic module mappings"


Summary:
Add intrinsic(fused) module mappings in quantize.py to enable mapping fused modules
in both QAT and post PTQ

Test Plan:
```
python test/test_qat.py
python test/test_quantization.py
```

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

jerryzh168 committed:
Update on "[quant] Add intrinsic module mappings"


Summary:
Add intrinsic(fused) module mappings in quantize.py to enable mapping fused modules
in both QAT and post PTQ

Test Plan:
```
python test/test_qat.py
python test/test_quantization.py
```

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

jerryzh168 committed:
Update on "[quant] Add intrinsic module mappings"


Summary:
Add intrinsic(fused) module mappings in quantize.py to enable mapping fused modules
in both QAT and post PTQ

Test Plan:
```
python test/test_qat.py
python test/test_quantization.py
```

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

raghuramank100 reviewed (approved):
Approved, with follow up needed to understand cause of mismatch of gradients for weights and batch norm gamma

jerryzh168 commented:
lint will is fixed by another pr

jerryzh168 committed:
Update on "[quant] Add intrinsic module mappings"


Summary:
Add intrinsic(fused) module mappings in quantize.py to enable mapping fused modules
in both QAT and post PTQ

Test Plan:
```
python test/test_qat.py
python test/test_quantization.py
```

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

Differential Revision: [D16820749](https://our.internmc.facebook.com/intern/diff/D16820749)

jerryzh168 committed:
Update on "[quant] Add intrinsic module mappings"


Summary:
Add intrinsic(fused) module mappings in quantize.py to enable mapping fused modules
in both QAT and post PTQ

Test Plan:
```
python test/test_qat.py
python test/test_quantization.py
```

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

Differential Revision: [D16820749](https://our.internmc.facebook.com/intern/diff/D16820749)

jerryzh168 committed:
Update on "[quant] Add intrinsic module mappings"


Summary:
Add intrinsic(fused) module mappings in quantize.py to enable mapping fused modules
in both QAT and post PTQ

Test Plan:
```
python test/test_qat.py
python test/test_quantization.py
```

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

Differential Revision: [D16820749](https://our.internmc.facebook.com/intern/diff/D16820749)

jerryzh168 committed:
Update on "[quant] Add intrinsic module mappings"


Summary:
Add intrinsic(fused) module mappings in quantize.py to enable mapping fused modules
in both QAT and post PTQ

Test Plan:
```
python test/test_qat.py
python test/test_quantization.py
```

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

Differential Revision: [D16820749](https://our.internmc.facebook.com/intern/diff/D16820749)
